sfge/src/test/java/com/salesforce/rules/dmlinloop/DmlInLoopRuleTest.java

```diff
@@ -0,0 +1,488 @@
+    }
+
+    /** The increment clause of a for loop repeats, which should be a violation */
+    @CsvSource({
+        "delete a",
+        "insert a",
+        "merge a a",
+        "undelete a",
+        "update a",
+        "upsert a",
+        "'Account b = [SELECT Id, Age FROM accounts WHERE Id = 3]'"
+    })
+    @ParameterizedTest(name = "{displayName}: {0}")
+    public void testForLoopIncrementStatement(String dmlStatement) {
+        // spotless:off
+        String[] sourceCode = {
+            "public class MyClass {\n" +
+                "private Integer i;\n" +
+                "MyClass() {\n" +
+                    "this.i = 0;\n" +
+                "}\n" +
+
+                "void incr() {\n" +
+                    "this.i++;\n" +
+                    "Account a = new Account();\n" +
+                    dmlStatement + ";\n" +
+                "}\n" +
+
+                "void foo() {\n" +
+                    "for (Integer i = 0; i < 4; incr()) {\n" +
+                    "}\n" +
+                "}\n" +
+            "}\n"
+        };
+        //spotless:on
+
+        assertViolations(
+                RULE,
+                sourceCode,
+                expect(
+                        9,
+                        new OccurrenceInfo(
+                                ASTConstants.NodeType.FOR_LOOP_STATEMENT, MY_CLASS, 12)));
+    }
+
+    /** Methods called within a loop containing DML are violations. */
+    @CsvSource({
+        "ForEachStatement, for (Integer i : myList), 'Account acc = [SELECT Id, Name, BillingCity FROM Accounts WHERE Id = 3 LIMIT 1];'",
+        "ForEachStatement, for (Integer i : myList), delete one;",
+        "ForEachStatement, for (Integer i : myList), insert one;",
+        "ForEachStatement, for (Integer i : myList), merge one one;",
+        "ForEachStatement, for (Integer i : myList), undelete one;",
+        "ForEachStatement, for (Integer i : myList), update one;",
+        "ForEachStatement, for (Integer i : myList), upsert one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), 'Account acc = [SELECT Id, Name, BillingCity FROM Accounts WHERE Id = 3 LIMIT 1];'",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), delete one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), insert one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), merge one one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), undelete one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), update one;",
+        "ForLoopStatement, for (Integer i = 0; i < 2; i++), upsert one;",
+        "WhileLoopStatement, while(true), 'Account acc = [SELECT Id, Name, BillingCity FROM Accounts WHERE Id = 3 LIMIT 1];'",
+        "WhileLoopStatement, while(true), delete one;",
+        "WhileLoopStatement, while(true), insert one;",
+        "WhileLoopStatement, while(true), merge one one;",
```

Comment:
I would recommend creating a scratch org so you can test whether this code actually compiles. I can walk you through that on Monday.

Reply:
Confirmed, it compiles.